extend.c

```diff
@@ -85,7 +85,7 @@ int mode_add_get(struct sockaddr_ina *dst, int m)
     }
     time(&t);
     if (t > val->time + params.cache_ttl) {
-        LOG(LOG_S, "time=%ld, now=%ld, ignore\n", val->time, t);
+        LOG(LOG_S, "time=%lld, now=%lld, ignore\n", (long long)val->time, (long long)t);
```

Comment:
This is still not a good fix for printing time_t.
See comments on https://stackoverflow.com/a/2792619

Reply:
I agree about possibly not good, but it is a _fix_ ;)

[Since C99, `long long` is the part of the standard](https://en.cppreference.com/w/c/language/arithmetic_types), and this project does require C99 (see Makefile). It means, this fix would fail only if `sizeof(long long) < sizeof(time_t)`, which is not possible by the standard: `long long` is the longest possible type in C99. and `[time_t` is a part of C99](https://en.cppreference.com/w/c/chrono/time_t).

So, for C99 it is _the fix_, and that discussion is pure speculation providing we speak about C99.

Reply:
The problem that `long long` by c standard should be at least 64bit in size. So it has no good use for 32bit platforms (Although modern linux tries to use 64bit for time_t even on 32bit platforms because of year 2038 problem, but there is still a big amount of old 32bit routers on old kernels).

Also i can suggest probably a better fix i can find, is to use `uintmax_t` like `printf("%ju", (uintmax_t)(clock_t)1);` from this answer: https://stackoverflow.com/a/17190680. 

time_t also can be floating type by the c standard, but i don't know usable systems with this.

Reply:
> still a big amount of old 32bit routers on old kernels

It is not a matter of the kernel. It's a matter of the toolchain used to build byedpi. My router is old, but the toolchain I used (musl) [uses 64-bit time_t for all platforms since version 1.2.0.](https://musl.libc.org/time64.html) It means, that log message without a fix may produce wrong output even on very old 32-bit platforms.

BTW if you care about old platforms, please consider to remove floating point operations from `main()`! Many CPUs in routers have no HW floating point 😄

Reply:
> without a fix may produce wrong output

Yes, this code without any fix is problematic. I agree that this need to be fixed. 

> My router is old, but the toolchain I used (musl) [uses 64-bit time_t for all platforms since version 1.2.0.](https://musl.libc.org/time64.html)

Musl uses compatibility layer to give you 64bit time_t when [still uses 32bit time_t under hood](https://git.musl-libc.org/cgit/musl/tree/compat/time32). 
While this works correctly for casting to `long long` it will still can introduce warnings when compiling for 32bit platform with native time_t.
However I just still think that `uintmax_t` will be better bet here. If it does not give any warnings at compiling with musl for 32bit and works correctly then it a better solution than `long long`. If it breaks musl then yes, `long long` is better.


> It is not a matter of the kernel.

Kernel actually matter for time64 syscalls support, but yes, this is not the scope of the issue.

> BTW if you care about old platforms, please consider to remove floating point operations from main()! Many CPUs in routers have no HW floating point 😄

There is only one usage of FP in the project, which can be pretty easily handled by software FP implementation. Can't complan.
```
$ grep -n -E "float|double" *.{c,h}
main.c:594:            float f = strtof(optarg, &end);
$ 
```

Reply:
> I just still think that `uintmax_t` will be better bet here

Then `intmax_t`, because, strictly speaking, `time_t` is signed value, to represent dates in the past.

OK, but what do you suggest for the format line then? I use "%**ll**d" for "`long long`", but which format is suitable for sure for `(u)intmax_t`?

Reply:
> OK, but what do you suggest for the format line then? I use "%lld" for "long long", but which format is suitable for sure for (u)intmax_t?

By c standard:
`%j` for `intmax_t` 
`%ju` for `uintmax_t`